File: code/ex992.java
// Sliding window

class Solution {
    public int subarraysWithKDistinct(int[] A, int K) {
        int n = A.length;
        // Elements in the interval which contains less than K distict elements
        Map<Integer, Integer> less = new HashMap<>();
        // Elements in the interval which contains less or equal than K distince element
        Map<Integer, Integer> lessEqual = new HashMap<>();
    
        
        int sl = 0, sle = 0;
        int res = 0;
        
        for (int hi = 0; hi < n; hi++) {
            less.put(A[hi], less.getOrDefault(A[hi], 0) + 1);
            lessEqual.put(A[hi], lessEqual.getOrDefault(A[hi], 0) + 1);
            
            // Move the starting pointer of two intervals to the right
            while (less.size() > K) {
                less.put(A[sl], less.get(A[sl]) - 1);
                if (less.get(A[sl]) == 0)    less.remove(A[sl]);
                sl++;
            }
            
            while (lessEqual.size() >= K) {
                lessEqual.put(A[sle], lessEqual.get(A[sle]) - 1);
                if (lessEqual.get(A[sle]) == 0)    lessEqual.remove(A[sle]);
                sle++;
            }
            
            res += sle - sl;
        }
        
        return res;
    }
}

File: code/ex992.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& A, int K) {
        int n = A.size(), res = 0;
        
        unordered_map<int, int> less;
        unordered_map<int, int> lessEqual;
        for(int l = 0, le = 0, i = 0; i < n; i++) {
            less[A[i]]++;
            lessEqual[A[i]]++;
            
            while (less.size() >= K) {
                if (--less[A[l]] == 0)    less.erase(A[l]);
                l++;
            }
            
            while (lessEqual.size() > K) {
                if (--lessEqual[A[le]] == 0)    lessEqual.erase(A[le]);
                le++;
            }
            
            res += l - le;
        }
        
        return res;
    }
};